diff --git a/taichi/backends/vulkan/runtime.h b/taichi/backends/vulkan/runtime.h
index 2b9dd826c..91b3a6357 100644
--- a/taichi/backends/vulkan/runtime.h
+++ b/taichi/backends/vulkan/runtime.h
@@ -103,13 +103,14 @@ class TI_DLL_EXPORT VkRuntime {
   void synchronize();
 
   Device *get_ti_device() const;
+  
+  void add_root_buffer(size_t root_buffer_size);
 
  private:
   friend class taichi::lang::vulkan::SNodeTreeManager;
 
   void init_nonroot_buffers();
 
-  void add_root_buffer(size_t root_buffer_size);
 
   Device *device_{nullptr};
   uint64_t *const host_result_buffer_;
